Add tmp directory before touching

yoongkang opened the pull request:
Addresses #20299

simi commented on railties/lib/rails/tasks/restart.rake:
```diff
@@ -1,4 +1,7 @@
 desc "Restart app by touching tmp/restart.txt"
 task :restart do
+  unless File.directory?('tmp')
+    system 'mkdir tmp'
```
It will be better to keep it all ruby and use [Dir::mkdir](http://ruby-doc.org/core-2.2.0/Dir.html#method-c-mkdir) to avoid various platform problems.

chancancode commented on railties/lib/rails/generators/rails/app/templates/gitignore:
```diff
@@ -17,5 +17,6 @@
 /log/*
 <% if keeps? -%>
 !/log/.keep
+!/tmp/.keep
```
We need to actually generate this file, grep the codebase for `empty_directory_with_keep_file` (probably need to add a test too)

chancancode commented on railties/lib/rails/tasks/restart.rake:
```diff
@@ -1,4 +1,5 @@
 desc "Restart app by touching tmp/restart.txt"
 task :restart do
+  Dir.mkdir('tmp') unless File.directory('tmp')
```
I think we can use `FileUtils.mkdir_p('tmp')` for this

chancancode replied:
Can you add a test for this? :)

https://github.com/rails/rails/blob/master/railties/test/application/rake/restart_test.rb

simi replied:
@chancancode why `mkdir_p` when there are not parent directories to create? Am I missing something?

chancancode replied:
If I understand it correctly, `mkdir_p` will work even if the folder is already created:

```
>> require 'fileutils'
=> true
>> Dir.exists?('zomg')
=> false
>> FileUtils.mkdir('zomg')
=> ["zomg"]
>> Dir.exists?('zomg')
=> true
>> FileUtils.mkdir('zomg')
Errno::EEXIST: File exists @ dir_s_mkdir - zomg
	from /Users/godfrey/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/fileutils.rb:252:in `mkdir'
	from /Users/godfrey/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/fileutils.rb:252:in `fu_mkdir'
	from /Users/godfrey/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/fileutils.rb:181:in `block in mkdir'
	from /Users/godfrey/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/fileutils.rb:180:in `each'
	from /Users/godfrey/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/fileutils.rb:180:in `mkdir'
	from (irb):5
	from /Users/godfrey/.rvm/rubies/ruby-2.2.0/bin/irb:11:in `<main>'
>> FileUtils.mkdir_p('zomg')
=> ["zomg"]
>> FileUtils.rmdir('zomg')
=> ["zomg"]
>> Dir.exists?('zomg')
=> false
>> FileUtils.mkdir_p('zomg')
=> ["zomg"]
>> Dir.exists?('zomg')
=> true
```

chancancode replied:
```
/tmp% rmdir zomg
/tmp% [[ -d zomg ]] && echo "ZOMG"
/tmp% mkdir zomg
/tmp% [[ -d zomg ]] && echo "ZOMG"
ZOMG
/tmp% mkdir zomg
mkdir: zomg: File exists
/tmp% mkdir -p zomg               
/tmp% [[ -d zomg ]] && echo "ZOMG"
ZOMG
/tmp% rmdir zomg
/tmp% mkdir -p zomg               
/tmp% [[ -d zomg ]] && echo "ZOMG"
ZOMG
```

chancancode commented:
I guess we ended up doing two things here, probably best to separate into two commits/changelog entries:

1. We changed the generator to `.keep` `/tmp` by default
2. We make the `rake restart` task not fail if `/tmp` isn't there

(These aren't the exact wording I'd use)

The first thing requires a CHANGELOG entry. The second thing doesn't need a CHANGELOG entry, because it's a bug fix for an unreleased feature (`rake restart` is only in master).

chancancode commented on railties/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+*   Create a `tmp` directory before touching if it does not exist, and keep the tmp directory
+    after cloning.
+    
+    *Yoong Kang Lim*, *Sunny Juneja*
```
:heart: very thoughtful of you to credit other contributors here :dancer: 

If you end your commit message(s) with `[Your Name, Another Name]` (on its own line), it'll register the commit for both of you on http://contributors.rubyonrails.org. See http://contributors.rubyonrails.org/faq#credit-several-people

chancancode commented:
Since you are amending your commits anyway, can you add "See #20299" (for the .keep one) and "Fixes #20299" (for the `rake restart` one) in the commit message/? It helps people reading it to lookup the context, and Github will also automatically close the issue when we merge this :smile:

chancancode commented on railties/lib/rails/generators/rails/app/templates/bin/setup:
```diff
@@ -25,5 +25,5 @@ chdir APP_ROOT do
   system 'ruby bin/rake log:clear tmp:clear'
```
Can you check if `tmp:clear` has the same bug?

yoongkang committed:
rake restart should create tmp folder

In restart.rake, the creation of tmp/restart.txt
would fail if the tmp folder does not exist in the
app. This is a problem because apps cloned using
git would not have the tmp folder, as the folder is
in .gitignore. This commit creates the tmp folder
if it does not exist.

Fixes #20299

[Yoong Kang Lim, Sunny Juneja]

yoongkang committed:
Create .keep file in tmp folder

The rake restart task creates tmp/restart.txt. This previously
errored out if the tmp folder does not exist, which is likely
given that the tmp folder is .gitignored by default. This
introduced problems when an app is cloned by git.

This commit makes sure that the generator creates a .keep file
when creating the tmp folder.

See #20299.

[Yoong Kang Lim, Sunny Juneja]

yoongkang committed:
Add test to ensure tmp:clear works when tmp is missing

See #20299.

yoongkang commented:
Thanks for the review @chancancode and @simi!

It should be ready for another round.

spastorino commented on railties/lib/rails/generators/rails/app/templates/bin/setup:
```diff
@@ -22,8 +22,9 @@ chdir APP_ROOT do
   system 'ruby bin/rake db:setup'
 
   puts "\n== Removing old logs and tempfiles =="
+  FileUtils.mkdir_p('tmp')
```
I wouldn't do this here, maybe rake tmp:clear should work if there's no tmp directory?

spastorino commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -141,6 +141,7 @@ def test
     end
 
     def tmp
+      empty_directory_with_keep_file "tmp"
```
Do we need this if we are already recreating the directory on each rake restart run?.

spastorino replied:
Or maybe we can leave this line as is and assume everywhere that we do have the tmp directory.

yoongkang replied:
I don't know. If we remove the creation of the tmp directory in rake restart, wouldn't cloning an older app still cause the original problem?

On the other hand is it the intended behavior to have the tmp folder exist after cloning? If we want it to exist, then we do need the .keep..

chancancode replied:
I consider the mkdir_p a workaround for a "bug" we have. Ideally we shouldn't have to do it if we have this, but we didn't have that for old apps so I guess we are stuck with the mkdir_p for now.

Actually that's probably my argument for `.keep`ing tmp. Without it, every script would have to do that check (and no one really do it, plenty of things just assume it exists). IMO, it should be like `logs`, you should always be able to write to it without having to worry about it not being there.

spastorino replied:
I wouldn't mind about older apps because the error is pretty obvious, but as you guys wish.

At least I'd make the restart task depend on another task that makes the mkdir_p and would call that task something like add_tmp_for_broken_cloned_apps or something like that

yoongkang replied:
:+1: added that task.

spastorino commented on railties/lib/rails/tasks/restart.rake:
```diff
@@ -1,4 +1,9 @@
 desc "Restart app by touching tmp/restart.txt"
 task :restart do
+  Rake.application.invoke_task(:add_tmp_if_missing)
```
Do not do it this way, make the restart task depend on it.

yoongkang replied:
my bad. changed it. thanks for the review.

chancancode commented:
Merged with some minor edits in d7cbdbe.

@spastorino @yoongkang I reverted the rake task dependency change – I might be wrong, but I *think* the intent is to have `rake restart` overridable to fit your setup, so that it eventually becomes a thing that other libraries can depend on. In that angle "create tmp" belongs to this specific implementation.

In general I agree that this is going to be a losing battle, and for the most part we should just assume the existence of `tmp` going forward. This in particular bugged me, because `bin/setup` is supposed to be the "first encounter" experience, and I would like that we make it as smooth as possible.

The error message you get is actually not very helpful: `touch: tmp/restart.txt: No such file or directory` (the real problem is `tmp` is not a folder, and also you need to know what `touch` does and what this `tmp/restart.txt` do), and I can see it being problematic for, e.g. designers who just need to setup the project and move on with making layout changes.

The maintenance overhead seems quite trivial, so it feels okay to go an extra mile (more like a meter) to make the onboarding experience a little better.

Happy to discuss and improve this more, but doesn't seem like they should block this, so I merged :smile:

chancancode commented:
@yoongkang thanks for working on this by the way :smile:

yoongkang commented:
Thanks for reviewing and merging! :+1:

spastorino commented:
:+1:
